### PR TITLE
DetailsList: adding slide animation to rows.

### PR DESCRIPTION
**Pull request by inateeg**
Description of changes
Adds a simple animation for individual cells when that cell's content updates. Side effect: entire list also slightly animates when it first renders, but it actually looks kind of nice.
When new content comes in, the current cell's content immediately disappears, and the new contents immediately start sliding in from the left.
The old, super-complex PR is here:
#5384
That implementation also animated the sliding out of the old content before animating the sliding in of the new content. Holding on to that old content caused huge amounts of extra complexity, so I cut it.
Focus areas to test
perf

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9432)

### CODE REVIEW

**Comment by aneeshack4** [comment]
Just putting the do not merge label on for now so no PRs get merged while we're locked down finishing up fabric 7.

**Comment by dzearing** [comment]
There are some conflicts to resolve, but overall this looks reasonable to me. Pinging @KevinTCoughlin  and @ThomasMichon on it.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts [hunk 1]
```diff
@@ -325,7 +334,7 @@ export const getStyles = (props: IDetailsRowStyleProps): IDetailsRowStyles => {
       }
     ],
 
-    cell: defaultCellStyles,
+    cell: [defaultCellStyles, enableUpdateAnimations && AnimationStyles.slideLeftIn40],
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts]: test in rtl. Is it going in the right direction? (pun intended)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 2]
```diff
@@ -52,9 +53,19 @@ export class DetailsRowFields extends React.PureComponent<IDetailsRowFieldsProps
               ? onRender(item, itemIndex, column)
               : getCellText(item, column);
 
+          // generate a key that auto-dirties when content changes, to force the container to re-render, to trigger animation
+          let key: number | string;
+          try {
+            const s = JSON.stringify(cellContentsRender);
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: woah. Please do not call JSON.stringify in this super duper hot spot perf place.

**Comment by KevinTCoughlin** [comment]
There were some recommendations when this feature was raised last month in a different PR, see https://github.com/OfficeDev/office-ui-fabric-react/pull/8936#discussion_r284352453. 

Can we incorporate that feedback and that which David has already provided? As-is this change will regress performance.

**Comment by aneeshack4** [comment]
Pinging @inateeg - it looks like there's some things to address from the reviewers.

**Comment by KevinTCoughlin** [comment]
I'm happy to look at this change in the profiler once merge conflicts are resolved so the example site deploys for test.

**Comment by inateeg** [comment]
Please stay tune, I am refactor the PR where we will get rid of stringigy call instead of supporting caller passing the valuekey.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 3]
```diff
@@ -43,25 +44,35 @@ export class DetailsRowFields extends React.PureComponent<IDetailsRowFieldsProps
...
             cellsByColumn && column.key in cellsByColumn
               ? cellsByColumn[column.key]
               : onRender
               ? onRender(item, itemIndex, column)
               : getCellText(item, column);
 
+          // generate a key that auto-dirties when content changes, to force the container to re-render, to trigger animation
+          let key: number | string;
+          try {
+            key = getValueKey ? getValueKey(item, itemIndex, column) : getCellText(item, column) + column.key + itemIndex;
```
Comment (soettl) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: That is potentially a very long string to be used as a key. `itemIndex` is unnecessary since the parent `<div>` contains only children of the same item. Also, `getValueKey` seems too specific to this use-case. There is no reason why this component needs to know about the semantics of the custom key being used.

Please do something along the lines of:
`let itemKey = this.props.getKey ? this.props.getKey(item, itemIndex, column, columnIndex) : columnIndex`

(similar how the key for a page is chosen in `List`).

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 4]
```diff
@@ -410,7 +420,7 @@ export interface IColumn {
   isResizable?: boolean;
 
   /**
-   * Determines if the column can render multi-line text.
+   * Determines if the column can render multi-line text.c
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: Please revert this change

**Comment by ecraig12345** [comment]
Given that the DetailsList's perf is already horrible and we have multiple people working on investigating/fixing that, we should **VERY** carefully consider whether these extra animations are actually necessary (especially on initial render) and find a way to measure their perf impact before checking in. Yes they're pretty, but is that really worth making the perf even worse?

**Comment by JasonGore** [comment]
Would it be worth doing another merge to pick up new perf-test? If it doesn't show any changes I could explore modifying perf-test to make sure it picks up these types of scenarios.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 3 above)
Comment (inateeg) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: the reason called getValueKey is emphasis key is made from value therefore value change key will change.  I can remove the fallback to just column.key + itemIndex to be unique and short.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts (hunk 4 above)
Comment (inateeg) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: thanks

**Comment by inateeg** [comment]
> Given that the DetailsList's perf is already horrible and we have multiple people working on investigating/fixing that, we should **VERY** carefully consider whether these extra animations are actually necessary (especially on initial render) and find a way to measure their perf impact before checking in. Yes they're pretty, but is that really worth making the perf even worse?

In theory the animation only happen if cell key changed therefore the very initial rendering should not has animation but subsequent update will have it.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts [hunk 5]
```diff
@@ -164,6 +167,15 @@ export interface IDetailsRowBaseProps extends Pick<IDetailsListProps, 'onRenderI
    */
   className?: string;
 
+  /**
+   * Whether to render shimmer
+   * @deprecated Use `ShimmeredDetailsList` instead: https://developer.microsoft.com/en-us/fabric#/components/detailslist/shimmer
+   */
+  shimmer?: boolean;
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts]: Is this PR adding a new prop and immediately deprecating it?

File: packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md [hunk 6]
```diff
@@ -3379,7 +3380,11 @@ export interface IDetailsListProps extends IBaseProps<IDetailsList>, IWithViewpo
     constrainMode?: ConstrainMode;
     disableSelectionZone?: boolean;
     dragDropEvents?: IDragDropEvents;
+    // @deprecated
+    enableShimmer?: boolean;
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md]: Another new prop that's being deprecated immediately? I can't find where this is being picked up from, either. I don't see a corresponding change in `DetailsList.types.ts`.

File: packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap [hunk 7]
```diff
@@ -5363,6 +5363,7 @@ exports[`DetailsList renders List with hidden checkboxes correctly 1`] = `
                                           {
                                             padding-right: 8px;
                                           }
+
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap]: This is weird. I wonder what's causing this.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts (hunk 5 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts]: This looks like the result of an incorrect merge. It seems that `6.0` was merged into this branch as this property is only available there and has since been removed. See: https://github.com/OfficeDev/office-ui-fabric-react/blob/1e1f4252a6278e2d4368fd1f3c84faebd2e09091/packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts#L171

File: packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md (hunk 6 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md]: Result of an incorrect merge with `6.0`

File: packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap (hunk 7 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap]: Could be IDE

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts (hunk 5 above)
Comment (inateeg) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts]: shimmer?:  is not part of my change.  should be removed ?

File: packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md (hunk 6 above)
Comment (inateeg) [reply packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md]: I should manually removed it since enableShimmer is not my change

Comment (JasonGore) [reply packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md]: You can't manually edit this file, it will just get regenerated based on PR contents. This file should **only** be updated via `npm run update-api`.

File: packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap (hunk 7 above)
Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap]: ?? It's a snapshot file, I'm not sure how IDE matters

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsList.test.tsx.snap]: If you have a plugin or somesuch and manually save the snapshot file prior to commit this can happen. Otherwise yea idk.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts (hunk 5 above)
Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.types.ts]: This is why it's appearing in the `api.md` file.

**Comment by statm** [comment]
@inateeg Looks like the newly added example needs to be added into a11y snapshot. Can you try `npm run update-snapshots` under apps/a11y-tests and push the updated snapshot file?

(Apology for the trouble. We are fixing the local build workflow so you don't have to do this in the future.)

**Comment by inateeg** [comment]
Bundle size auditor seems wrong.  It posted exact the same size on both columns master and PR ?  Can someone look at this?

**Comment by aftab-hassan** [comment]
> Bundle size auditor seems wrong. It posted exact the same size on both columns master and PR ? Can someone look at this?

Looks like master branch is broken on this one. There is no baseline to compare against.

**Comment by inateeg** [comment]
who can look into why OfficeDev master bundle size baseline comparison seems all wrong?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 3 above)
Comment (soettl) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: `DetailsRowFields` should be agnostic of what type of key is going to be passed. What if some other component wants to use `DetailsRowFields`, but not pass a value as a key, but something else? I don't see any reason why we would name this new prop after this specific use case of a value changing.

Also, why is the previously used `columnIndex` not sufficient in case no `getCellKey` is passed?

**Comment by aftab-hassan** [comment]
> who can look into why OfficeDev master bundle size baseline comparison seems all wrong?

discussed offline with @inateeg. The new commit https://github.com/OfficeDev/office-ui-fabric-react/pull/9432/commits/2a22ec80922f88cbddea051271ea06a88f2069ca made a couple minutes ago should update bundle size information

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 8]
```diff
@@ -43,25 +44,35 @@ export class DetailsRowFields extends React.PureComponent<IDetailsRowFieldsProps
...
           const cellContentsRender =
             cellsByColumn && column.key in cellsByColumn
               ? cellsByColumn[column.key]
               : onRender
               ? onRender(item, itemIndex, column)
               : getCellText(item, column);
 
+          // generate a key that auto-dirties when content changes, to force the container to re-render, to trigger animation
+          let key: number | string;
+          try {
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: why is a try catch needed here?

File: packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Animation.Example.tsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    alert(`Item invoked: ${item.name}`);
+  };
+
+  private _getValueKey(item: IDetailsListAnimationExampleItem, index: number, column: IColumn): string {
+    const key = item && column && column.fieldName ? item[column.fieldName as keyof IDetailsListAnimationExampleItem] : column.key + index;
+    return key.toString();
+  }
+
+  private _getCellValueKey(item: IDetailsListAnimationExampleItem, index: number, column: IColumn): string {
+    const key = item && column && column.fieldName ? item[column.fieldName as keyof IDetailsListAnimationExampleItem] : column.key + index;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Animation.Example.tsx]: if column is null you will nullref

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts [hunk 10]
```diff
@@ -14,6 +15,7 @@ const GlobalClassNames = {
   root: 'ms-DetailsRow',
   compact: 'ms-DetailsList--Compact', // TODO: in Fabric 7.0 lowercase the 'Compact' for consistency across other components.
   cell: 'ms-DetailsRow-cell',
+  cellAnimation: 'ms-DetailsRow-cellAnimiation',
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts]: `cellAnimiation` has a typo.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 8 above)
Comment (inateeg) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: original thinking must be that if user passed in callback caused exception should not impact the rendering of list?

Comment (inateeg) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: perhaps we should trust this just as trust the rest of callback?!

**Comment by KevinTCoughlin** [comment]
Taking a look now that CI is passing

**Comment by inateeg** [comment]
> perf regression.

removed stringify and perf test seems accetable.

**Comment by inateeg** [comment]
perf test result looks reasonable.

**Comment by dzearing** [comment]
@inateeg I will take another look

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts [hunk 11]
```diff
@@ -330,8 +333,8 @@ export const getStyles = (props: IDetailsRowStyleProps): IDetailsRowStyles => {
       }
     ],
 
-    cell: defaultCellStyles,
-
+    cell: [defaultCellStyles],
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts]:
```suggestion
    cell: defaultCellStyles,
```

Shouldn't need to convert this to an array since it is a single value, please remove

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts [hunk 12]
```diff
@@ -330,8 +333,8 @@ export const getStyles = (props: IDetailsRowStyleProps): IDetailsRowStyles => {
...
+    cellAnimation: [enableUpdateAnimations && AnimationStyles.slideLeftIn40],
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.styles.ts]:
```suggestion
    cellAnimation: enableUpdateAnimations && AnimationStyles.slideLeftIn40,
```

Should work

File: packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Animation.Example.tsx [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+      'aliquip ex ea commodo consequat duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore ' +
+      'eu fugiat nulla pariatur excepteur sint occaecat cupidatat non proident sunt in culpa qui officia deserunt '
+    ).split(' ');
+    let loremIndex = 0;
+    function _lorem(wordCount: number): string {
+      const startIndex = loremIndex + wordCount > LOREM_IPSUM.length ? 0 : loremIndex;
+      loremIndex = startIndex + wordCount;
+      return LOREM_IPSUM.slice(startIndex, loremIndex).join(' ');
+    }
+    setInterval(() => {
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Animation.Example.tsx]: This timer needs to be cleared on `componentWillUnmount`

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Animation.Example.tsx]: You might also want to defer setting it until `componentDidMount` instead of the constructor

File: packages/office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.List.Example.tsx [hunk 14]
```diff
@@ -26,7 +26,7 @@ const COLUMNS: IColumn[] = [
     onRender: item => <Link href={item.url}>{item.url}</Link>
   },
   {
-    key: 'link',
+    key: 'link2',
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.List.Example.tsx]: Revert

File: packages/office-ui-fabric-react/src/components/__snapshots__/TextField.AutoComplete.Example.tsx.shot [hunk 15]
```diff
@@ -62,37 +62,6 @@ exports[`Component Examples renders TextField.AutoComplete.Example.tsx correctly
           ms-TextField-wrapper
 
     >
-<<<<<<< HEAD
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/__snapshots__/TextField.AutoComplete.Example.tsx.shot]: I don't see these merge conflict markers committed in master. Not sure what happened here.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 16]
```diff
@@ -419,6 +429,11 @@ export interface IColumn {
    */
   onRender?: (item?: any, index?: number, column?: IColumn) => any;
 
+  /**
+   * If set, parent getCellValueKey will return this value.
+   */
+  getValueKey?: (item?: any, index?: number, column?: IColumn) => string;
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: Are both `getCellValueKey` and `getValueKey` necessary? They seem redundant in signature and implementation in the example below.

Can we collapse it to only one new method to implement?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.test.tsx [hunk 17]
```diff
@@ -312,6 +312,14 @@ describe('DetailsList', () => {
       );
     };
 
+    const getCellValueKey = (item: any, index: number, column: IColumn) => {
+      let valueKey = item && column && column.fieldName ? item[column.fieldName] : '';
+      if (valueKey === null || valueKey === undefined) {
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.test.tsx]: Perhaps also verify `valueKey` is not an empty string else it will log warnings for duplicate keys.

File: packages/experiments/src/components/SelectedItemsList/SelectedItemsList.tsx [hunk 18]
```diff
@@ -128,4 +128,4 @@ const _SelectedItemsList = <TItem extends BaseSelectedItem>(
 // Typescript only respects unifying a generic type with a generic const _function_ of the same name for function types.
 // In order to satisfy the type checker, here we lie about the type of the const so that it is still a generic function.
 export type SelectedItemsList<TItem extends BaseSelectedItem> = React.Component<ISelectedItemsListProps<TItem>>;
-export const SelectedItemsList = React.forwardRef(_SelectedItemsList) as (typeof _SelectedItemsList);
+export const SelectedItemsList = React.forwardRef(_SelectedItemsList) as typeof _SelectedItemsList;
```
Comment (KevinTCoughlin) [inline packages/experiments/src/components/SelectedItemsList/SelectedItemsList.tsx]: Revert changes to this file and you can then remove this change file: https://github.com/OfficeDev/office-ui-fabric-react/pull/9432/files#diff-3fdb5afc2d18ec3c9108edf97b679320

File: common/changes/@uifabric/experiments/inateeg-0611list_2019-06-26-19-24.json [hunk 19]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "",
+      "packageName": "@uifabric/experiments",
```
Comment (KevinTCoughlin) [inline common/changes/@uifabric/experiments/inateeg-0611list_2019-06-26-19-24.json]: This change file shouldn't be needed, see https://github.com/OfficeDev/office-ui-fabric-react/pull/9432/files#diff-a4c06458fab060a6f0be01f6290a290a

**Comment by KevinTCoughlin** [comment]
FYI it looks like as-is, this feature will cause React to update all rows and cells in a List (updates are depicted by the cool-to-warm borders in the below screencapture).

It would be ideal if the animation only caused the actual cell + row combination to update:

list_animation

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts (hunk 16 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: Spoke offline, this is required for per column overrides similar to `IColumn#onRender`

File: packages/office-ui-fabric-react/src/components/__snapshots__/TextField.Placeholder.Example.tsx.shot [hunk 20]
```diff
@@ -207,17 +207,8 @@ exports[`Component Examples renders TextField.Placeholder.Example.tsx correctly
         className=
             ms-Label
             {
-<<<<<<< HEAD
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/__snapshots__/TextField.Placeholder.Example.tsx.shot]: FYI - this diff is correct.. these snapshots were merged with conflicts:

https://github.com/OfficeDev/office-ui-fabric-react/blob/044e13c2c8c391885693bbaa493e975fab35f3ea/packages/office-ui-fabric-react/src/components/__snapshots__/TextField.Placeholder.Example.tsx.shot#L210
